File: Excel/429 Challenge.py
import pandas as pd
import numpy as np
import math

input = pd.read_excel("429 Pythagorean Quadruples.xlsx", sheet_name="Sheet1", usecols="A", nrows=10)
test = pd.read_excel("429 Pythagorean Quadruples.xlsx", sheet_name="Sheet1", usecols="B", nrows=10) 

def find_quadr_solution(sides):
    sides = sides.replace("X", "0")

    numbers = np.array([float(num) for num in sides.split(", ") if num != 'NA'])
    numbers = numbers[numbers != 0]
    
    # Calculate missing side assuming d is missing
    missing1 = np.sqrt(np.sum(numbers**2))
    
    # Calculate missing side assuming a, b, or c is missing
    pot_d = np.max(numbers)
    others = numbers[numbers != pot_d]
    missing2 = np.sqrt(pot_d**2 - np.sum(others**2))
    
    if missing1 == np.floor(missing1):
        missing = missing1
    elif missing2 == np.floor(missing2):
        missing = missing2
    else:
        missing = np.nan 

    missing = np.int64(missing)
    return missing 

result = input.copy()
result["r"] = result["Number"].apply(find_quadr_solution)

print(result["r"].equals(test["Answer Expected"])) # True

File: Excel/400-499/429/429 Challenge.R
library(tidyverse)
library(readxl)

input = read_excel("Excel/429 Pythagorean Quadruples.xlsx", range = "A1:A10")
test  = read_excel("Excel/429 Pythagorean Quadruples.xlsx", range = "B1:B10")

find_quadr_solution = function(sides) {
  numbers = str_split(sides, ", ")[[1]] %>% 
    as.numeric() %>%
    na.omit()
  
  missing1 = sqrt(sum(numbers^2)) # if d side is missing
  
  pot_d = max(numbers)
  others = numbers[numbers != pot_d]
  missing2 = sqrt(pot_d^2 - sum(others^2)) # if a, b or c side is missing

  if (missing1 == floor(missing1)) {
    missing = missing1
  } else if (missing2 == floor(missing2)) {
    missing = missing2
  } else {
    missing = NA
  }

  return(missing)
}

result = input %>%
  mutate(r = map_dbl(Number, find_quadr_solution)) 

identical(result$r, test$`Answer Expected`)
# [1] TRUE
